packages/ckeditor5-engine/src/conversion/downcasthelpers.js

```diff
@@ -1670,6 +1896,247 @@ function prepareDescriptor( highlightDescriptor, data, conversionApi ) {
+		 * the children elements would not be assigned to any of the slots).
+		 *
+		 * @error conversion-slot-filter-incomplete
+		 * @param {module:engine/model/element~Element} element The element of which children would not be properly
+		 * allocated to multiple slots.
+		 */
+		throw new CKEditorError( 'conversion-slot-filter-incomplete', conversionApi.dispatcher, { element } );
+	}
+}
+
+// Fill slots with appropriate view elements.
+//
+// @param {module:engine/view/element~Element} viewElement
+// @param {Map.<module:engine/view/element~Element,Array.<module:engine/model/node~Node>>} slotsMap
+// @param {module:engine/conversion/downcastdispatcher~DowncastConversionApi} conversionApi
+// @param {Object} options
+// @param {Boolean} [options.reconversion]
+function fillSlots( viewElement, slotsMap, conversionApi, options ) {
+	// Set temporary position mapping to redirect child view elements into a proper slots.
+	conversionApi.mapper.on( 'modelToViewPosition', toViewPositionMapping, { priority: 'highest' } );
+
+	let currentSlot = null;
+	let currentSlotNodes = null;
+
+	// Fill slots with nested view nodes.
+	for ( [ currentSlot, currentSlotNodes ] of slotsMap ) {
+		reinsertNodes( viewElement, currentSlotNodes, conversionApi, options );
+
+		conversionApi.writer.move(
+			conversionApi.writer.createRangeIn( currentSlot ),
+			conversionApi.writer.createPositionBefore( currentSlot )
+		);
+		conversionApi.writer.remove( currentSlot );
+	}
+
+	conversionApi.mapper.off( 'modelToViewPosition', toViewPositionMapping );
+
+	function toViewPositionMapping( evt, data ) {
+		const element = data.modelPosition.nodeAfter;
+
+		// Find the proper offset within the slot.
+		const index = currentSlotNodes.indexOf( element );
+
+		if ( index < 0 ) {
+			return;
+		}
+
+		data.viewPosition = data.mapper.findPositionIn( currentSlot, index );
+	}
+}
+
+// Inserts view representation of `nodes` into the `viewElement` either by bringing back just removed view nodes
+// or by triggering conversion for them.
+//
+// @param {module:engine/view/element~Element} viewElement
+// @param {Iterable.<module:engine/model/element~Element>} modelNodes
+// @param {module:engine/conversion/downcastdispatcher~DowncastConversionApi} conversionApi
+// @param {Object} options
+// @param {Boolean} [options.reconversion]
+function reinsertNodes( viewElement, modelNodes, conversionApi, options ) {
+	const { writer, mapper } = conversionApi;
+
+	// Fill with nested view nodes.
+	for ( const modelChildNode of modelNodes ) {
+		const viewChildNode = mapper.toViewElement( modelChildNode );
+
+		if ( options.reconversion && viewChildNode && viewChildNode.root != viewElement.root ) {
+			writer.move(
+				writer.createRangeOn( viewChildNode ),
+				mapper.toViewPosition( ModelPosition._createBefore( modelChildNode ) )
+			);
+		} else {
+			conversionApi.convertInsert( ModelRange._createOn( modelChildNode ) );
+		}
```

Comment:
Clarification needed.

1.  Are there two modes because this function is sometimes used for the "first insertion" and sometimes for reconversion?
2.  Do we check `viewChildNode` when there's `reconversion` flag because of text nodes?
3.  Why do we check roots?

Reply:
1.  This is for cases when reconversion is triggered by adding/removing child nodes. The old view elements are restored and only the added child is converted.
2.  We want to restore old view elements only for reconversion and not while the initial insert.
3.  To make sure that that view element is not already added to the view structure.